### ISSUE TITLE
Support learning lemmas that introduce new variables

### ISSUE THREAD

**Issue by jmgrosen** [bedrocksystems/cpp2v#170]
I want to add a learning lemma like this:
```coq
forall v l P Q,
  (forall p,
    v = Vptr p ->
    _at (_eq v) (Inv_List l) ** P |-- Q) ->
  _at (_eq v) (Inv_List l) ** P |-- Q.
```

However, if this is added to the learning database right now, `learn_here` will loop forever, because `learn` sees new hypotheses each time -- first `v = Vptr p0`, then `v = Vptr p1`, `v = Vptr p2`, and so on. Quantifying `p` in different ways only serves to delay the problem.

[This is solved for the case of `refine_tprim_ptr`](https://github.com/bedrocksystems/cpp2v/blob/186addf3b63c4960a572dbc26d698e042ab1b055/theories/Auto/vc.v#L205-L209) by
1. checking if the introduced hypothesis is of the form `Vptr _ = Vptr _` and failing if so, combined with immediately destructing any equalities that *are* introduced
2. changing the goal in such a way that the lemma won't apply again

(1) would work for this special case, too, but it's not clear to me how to generalize it well; (2) doesn't seem to immediately apply to this case.

**Comment by jmgrosen** [comment]
@gmalecha

**Comment by gmalecha** [comment]
This is meant to be addressed by https://github.com/bedrocksystems/cpp2v/blob/master/theories/Auto/Discharge.v#L588 . If it doesn't work, then we should modify that.

**Comment by jmgrosen** [comment]
I agree that is where it should probably be fixed; however, I don't know how. Here's an illustration of the issue:

- Assume the lemma I wrote above is added to the learning database.
- When `learn_here` applies it, we get a goal like `forall p, v = Vptr p -> _at (_eq v) (Inv_List l) ** P |-- Q`.
- When `learn` is called, it first intros `p` because it always intros on dependent products. We now have a context hypothesis `p0 : ptr`.
- When `learn` recurses, it sees the non-dependent implication `v = Vptr p0` and calls `doesn't_know` on it to see if we already know it. None of the failing guards match, so it succeeds and `learn` adds `H0 : v = Vptr p0` to the context.
- Now `learn_here` again applies the same lemma, resulting in the goal `forall p, v = Vptr p -> _at (_eq v) (Inv_List l) ** P |-- Q`.
- When `learn` is called, it first intros `p` because it always intros on dependent products. We now have a context hypothesis `p1 : ptr`.
- When `learn` recurses, it sees the non-dependent implication `v = Vptr p1`, and calls `doesn't_know` on it. Here, perhaps, we should find that we already know it, but we don't -- we have `H0 : v = Vptr p0` in our context, but nothing with type `v = Vptr p1`.
- We are now stuck in a loop.

One semi-hack that could fix this is if we instead changed the type of the lemma to
```coq
forall v l P Q,
  (forall p,
    v = Vptr p ->
    _at (_eq (Vptr p)) (Inv_List l) ** P |-- Q) ->
  _at (_eq v) (Inv_List l) ** P |-- Q.
```
(notice the transformed goal no longer mentions `v`) and then, in `doesn't_know`, we checked to make sure that the fact doesn't have type `Vptr _ = Vptr _`. But this doesn't seem like a very general solution.

**Comment by gmalecha** [comment]
On option would be to do something like: "don't learn the goal of it is provable by `tactic`". And that could be something like `congruence` or `assumption`.

**Comment by jmgrosen** [comment]
How does that help in this scenario? `assumption` is (part of) what `doesn't_know` is already doing, and I don't see how `congruence` improves on the situation.

**Comment by jmgrosen** [comment]
Perhaps it makes more sense to hack on in person sometime.

**Comment by gmalecha** [comment]
I'll look into it tonight and we can look at it tomorrow in person.

**Comment by gmalecha** [comment]
I see what the issue is. Maybe we need something like:

```coq
Definition require_not_provable (P Q : Prop) : Prop := Q.
```

When we are intro-ing in `learn`, we see if we have a `require_not_provable P Q`, and if we do, we test to see if `P` is provable. If it is, then we fail, otherwise, we essentially `red` it and drop the `P` and continue.
